### PR TITLE
[IMP] hr_expense: expense report

### PR DESCRIPTION
**Pull request by malb-odoo**
This pr aim to improve the pdf of the expense report. we have changed the design
 of the header of the report, the table and also added an annex with the
 different attachment.

task: 3443413

---
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: addons/hr_expense/models/ir_actions_report.py [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+                    for attachment in attachments:
+                        if attachment.mimetype == 'application/pdf':
+                            attachment_stream = pdf.to_pdf_stream(attachment)
+                            attachment_pdf = OdooPdfFileReader(attachment_stream, strict=False)
+                            output_pdf.appendPagesFromReader(attachment_pdf)
+                        else:
+                            data['attachment'] = attachment
+                            attachment_img = self._render_qweb_pdf_prepare_streams('hr_expense.report_expense_sheet_img', data, res_ids=res_ids)
+                            attachment_img_stream = OdooPdfFileReader(attachment_img[expense_sheet.id]['stream'], strict=False)
+                            output_pdf.appendPagesFromReader(attachment_img_stream)
```
Comment (clbr-odoo) [inline addons/hr_expense/models/ir_actions_report.py:39]: Hum there's something strange in your variable names here that makes the logic harder to understand imo.
What do you think of this ?
```suggestion
                        if attachment.mimetype == 'application/pdf':
                            attachment_stream = pdf.to_pdf_stream(attachment)
                        else:
                            data['attachment'] = attachment
                            attachment_prep_stream = self._render_qweb_pdf_prepare_streams('hr_expense.report_expense_sheet_img', data, res_ids=res_ids)
                            attachment_stream = attachment_prep_stream[expense_sheet.id]['stream']

                        attachment_reader = OdooPdfFileReader(attachment_stream, strict=False)
                        output_pdf.appendPagesFromReader(attachment_reader)
```

File: addons/hr_expense/models/ir_actions_report.py [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+
+                        new_pdf_stream = io.BytesIO()
+                        output_pdf.write(new_pdf_stream)
+                        res[expense_sheet.id]['stream'] = new_pdf_stream
+                        stream.close()
```
Comment (clbr-odoo) [inline addons/hr_expense/models/ir_actions_report.py:44]: Shouldn't this be done one tab left ? At the same level as `stream = ...` (where you get the stream) ?
Maybe this portion is worth a comment to indicate what you're doing. :thinking:

**Review by clbr-odoo — CHANGES_REQUESTED** [review]
Very nice task, few comments/questionmarks :slightly_smiling_face:  !
And side note that it won't work for other documents types than images and pdfs, but it was already the case before and it does not throws ugly errors, it just displays an empty page so fine by me :+1:

**Review by clbr-odoo — CHANGES_REQUESTED** [review]

File: addons/hr_expense/models/ir_actions_report.py [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+    def _render_qweb_pdf_prepare_streams(self, report_ref, data, res_ids=None):
+        # OVERRIDE
+        res = super()._render_qweb_pdf_prepare_streams(report_ref, data, res_ids)
+        if not res_ids:
+            return res
+        report = self._get_report(report_ref)
+        if report.report_name == 'hr_expense.report_expense_sheet':
+            expense_sheets = self.env['hr.expense.sheet'].browse(res_ids)
+            if expense_sheets:
+                streams_to_append = {}
```
[the remark this answers is not in the document]
Comment (malb-odoo) [reply addons/hr_expense/models/ir_actions_report.py:19]: ooopss

**Review by oco-odoo — CHANGES_REQUESTED** [review]

File: addons/hr_expense/models/ir_actions_report.py [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+                    attachments = self.env['ir.attachment'].search([('res_id', 'in', expense_sheet.expense_line_ids.ids), ('res_model', '=', 'hr.expense')])
+                    expense_report = OdooPdfFileReader(stream, strict=False)
+                    output_pdf = OdooPdfFileWriter()
+                    output_pdf.appendPagesFromReader(expense_report)
+                    for attachment in attachments:
+                        if attachment.mimetype == 'application/pdf':
+                            attachment_stream = pdf.to_pdf_stream(attachment)
+                        else:
+                            data['attachment'] = attachment
+                            attachment_prep_stream = self._render_qweb_pdf_prepare_streams(
```
[the remark this answers is not in the document]
Comment (malb-odoo) [reply addons/hr_expense/models/ir_actions_report.py:31]: i will put a comment to explain :laughing:

**Review by oco-odoo — APPROVED** [review]
@robodoo r+

**Pull request closed without merging**